tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc

```diff
@@ -174,6 +175,116 @@ DECL_CONVERT_OP(FakeQuant);
 
 #undef DECL_CONVERT_OP
 
+LogicalResult ConvertTFLGeluOp::matchAndRewrite(
+    Operation* op, PatternRewriter& rewriter) const {
+  auto tfl_gelu_op = cast<TFL::GeluOp>(op);
+  Location loc = op->getLoc();
+
+  Value input = tfl_gelu_op.input();
+  ShapedType input_type = input.getType().dyn_cast<ShapedType>();
+  ShapedType output_type =
+      tfl_gelu_op.getResult().getType().dyn_cast<ShapedType>();
+  // Not a ranked tensor output
+  if (!input_type || !output_type) return failure();
+
+  auto input_shape = input_type.getShape();
+  auto output_shape = output_type.getShape();
+
+  bool input_is_qtype =
+      input_type.getElementType().isa<mlir::quant::UniformQuantizedType>();
+  bool output_is_qtype =
+      output_type.getElementType().isa<mlir::quant::UniformQuantizedType>();
+
+  if (input_is_qtype != output_is_qtype) {
+    return rewriter.notifyMatchFailure(
+        op,
+        "ConvertTFLGeluOp: input/output tensor should be all quantized or all "
+        "floating-point.");
+  }
+
+  if (output_is_qtype) {
+    UniformQuantizedType input_qtype =
+        input_type.getElementType()
+            .dyn_cast<mlir::quant::UniformQuantizedType>();
```

Comment:
You can just cast here as you used isa above. Alternatively, you can dyn_cast below and then just use in here if not null. That is more common. It would require changing the expression in line 198 unless this operation is constrained to same input and result type even for quantized.

Reply:
Great explanation. Thanks.
I adopt to dyn_cast above and then check if tensors are quantized by evaluating whether the type is null. And yes, the expression is required to change as you mentioned.